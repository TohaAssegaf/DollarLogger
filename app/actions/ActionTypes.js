import {
    Payment
} from '/app/store/state/PaymentState'

// Set budget total.
export const SET_BUDGET_TOTAL_REQUEST = 'SET_BUDGET_TOTAL_REQUEST'
export type SetBudgetTotalRequestAction = {
    type: SET_BUDGET_TOTAL_REQUEST
}
export const SET_BUDGET_TOTAL_SUCCESS = 'SET_BUDGET_TOTAL_SUCCESS'
export type SetBudgetTotalSuccessAction = {
    type: SET_BUDGET_TOTAL_SUCCESS,
    total: number
}
export const SET_BUDGET_TOTAL_FAILURE = 'SET_BUDGET_TOTAL_FAILURE'
export type SetBudgetTotalFailureAction = {
    type: SET_BUDGET_TOTAL_FAILURE,
    errorMessage: string
}

// Get budget total.
export const GET_BUDGET_TOTAL_REQUEST = 'GET_BUDGET_TOTAL_REQUEST'
export type GetBudgetTotalRequestAction = {
    type: GET_BUDGET_TOTAL_REQUEST
}
export const GET_BUDGET_TOTAL_SUCCESS = 'GET_BUDGET_TOTAL_SUCCESS'
export type GetBudgetTotalSuccessAction = {
    type: GET_BUDGET_TOTAL_SUCCESS,
    total: number
}
export const GET_BUDGET_TOTAL_FAILURE = 'GET_BUDGET_TOTAL_FAILURE'
export type GetBudgetTotalFailureAction = {
    type: GET_BUDGET_TOTAL_FAILURE,
    errorMessage: string
}

// Create payment.
export const CREATE_PAYMENT_REQUEST = 'CREATE_PAYMENT_REQUEST'
export type CreatePaymentRequestAction = {
    type: CREATE_PAYMENT_REQUEST
}
export const CREATE_PAYMENT_SUCCESS = 'CREATE_PAYMENT_SUCCESS'
export type CreatePaymentSuccessAction = {
    type: CREATE_PAYMENT_SUCCESS,
    payment: Payment
}
export const CREATE_PAYMENT_FAILURE = 'CREATE_PAYMENT_FAILURE'
export type CreatePaymentFailureAction = {
    type: CREATE_PAYMENT_FAILURE,
    errorMessage: string
}

export type Action = |
    SetBudgetTotalRequestAction |
    SetBudgetTotalSuccessAction |
    SetBudgetTotalFailureAction |
    GetBudgetTotalRequestAction |
    GetBudgetTotalSuccessAction |
    GetBudgetTotalFailureAction |
    CreatePaymentRequestAction |
    CreatePaymentSuccessAction |
    CreatePaymentFailureAction