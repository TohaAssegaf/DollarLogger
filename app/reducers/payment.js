import {
  GET_PAYMENTS_REQUEST,
  GET_PAYMENTS_SUCCESS,
  GET_PAYMENTS_FAILURE,
  Action
} from '~/app/actions/ActionTypes'
import {
  Payment,
  PaymentState
} from '~/app/store/state/PaymentState'

const initialState: PaymentState = {
  payments: [],
  isLoading: true,
  errorMessage: '',
}

export default function (state: PaymentState = initialState, action: Action): PaymentState {
  switch (action.type) {
  case GET_PAYMENTS_REQUEST:
    return getPaymentsRequest(state, action)
  case GET_PAYMENTS_SUCCESS:
    return getPaymentsSuccess(state, action)
  case GET_PAYMENTS_FAILURE:
    return getPaymentsFailure(state, action)
  default:
    return state
  }
}

function getPaymentsRequest(
  state: PaymentState, action: GetPaymentsRequestAction): PaymentState {
  return Object.assign({}, state, {
    isLoading: true,
    errorMessage: ""
  })
}

function getPaymentsSuccess(
  state: PaymentState, action: GetPaymentsSuccessAction): PaymentState {
  let payments = action.payments.slice()
  return Object.assign({}, state, {
    payments,
    isLoading: false,
  })
}

function getPaymentsFailure(
  state: PaymentState, action: GetPaymentsFailureAction): PaymentState {
  return Object.assign({}, state, {
    errorMessage: action.errorMessage,
    isLoading: false,
  })
}
