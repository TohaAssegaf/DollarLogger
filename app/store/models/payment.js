import { PAYMENTS_ASYNC_STORAGE_KEY } from '~/app/config/storage'
import { AsyncStorage } from 'react-native'

export function getPayments() {
  return AsyncStorage.getItem(PAYMENTS_ASYNC_STORAGE_KEY)
    .then(payments => {
      if (!payments) {
        return []
      }
      return JSON.parse(payments).map(payment => parsePayment(payment))
    })
}

/** Need to parse all date fields into actual Date objects instead of numbers. */
function parsePayment(payment) {
  let paymentContributions = payment.paymentContributions.map(
    paymentContribution =>
      Object.assign({}, paymentContribution, { date: new Date(paymentContribution.date) }))
  return Object.assign({}, payment, { date: new Date(payment.date), paymentContributions })
}

export function addPayment(total: number, name: string, date: Date) {
  return getPayments().then(payments => {
    // TODO(renzobautista): Separate ID generation into a new class so it can be mocked
    const id = new Date().getTime()
    const payment = {
      id,
      total,
      name,
      date,
      paymentContributions: [
        {
          total,
          date,
        }
      ],
    }
    payments.push(payment)
    return AsyncStorage.setItem(PAYMENTS_ASYNC_STORAGE_KEY, JSON.stringify(payments))
      .then(() => payments)
  })
}

export function updatePayment(payment: Payment) {
  return getPayments().then(payments => {
    const updatedPayments = payments.map(
      storedPayment => storedPayment.id === payment.id ? payment : storedPayment)
    return AsyncStorage.setItem(PAYMENTS_ASYNC_STORAGE_KEY, JSON.stringify(updatedPayments))
      .then(() => updatedPayments)
  })
}

export function deletePayment(id: number) {
  return getPayments().then(payments => {
    const updatedPayments = payments.filter(payment => payment.id != id)
    return AsyncStorage.setItem(PAYMENTS_ASYNC_STORAGE_KEY, JSON.stringify(updatedPayments))
      .then(() => updatedPayments)
  })
}
