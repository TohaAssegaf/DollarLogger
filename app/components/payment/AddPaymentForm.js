import BasePaymentForm from './BasePaymentForm'
import actions from '~/app/actions'
import PaymentBuilder from '~/app/lib/PaymentBuilder'
import React from 'react'
import { connect } from 'react-redux'

class AddPaymentForm extends React.Component {
  createPayment(total, name, date, splitCount) {
    this.props.createPayment(
      new PaymentBuilder()
        .setName(name)
        .setTotal(total)
        .setDate(date)
        .setSplitCount(splitCount)
        .build())

    // Temporary hack. This should check state with a componentDidUpdate for write to be complete.
    this.props.navigation.goBack()
  }

  render() {
    return (
      <BasePaymentForm
        onSubmit={
          (total, name, date, splitCount) => this.createPayment(total, name, date, splitCount)}
      />
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    createPayment: (payment) => {
      dispatch(actions.createPayment(payment))
    }
  }
}

export default connect((state) => ({}), mapDispatchToProps)(AddPaymentForm)
