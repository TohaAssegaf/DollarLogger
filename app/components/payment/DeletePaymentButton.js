import actions from '/app/actions'
import { DANGER_BUTTON_COLOR } from '/app/config/colors'
import React from 'react'
import { Alert, Button } from 'react-native'
import { connect } from 'react-redux'

class DeletePaymentButton extends React.Component {
  deletePayment() {
    Alert.alert(
      'Delete payment?',
      "This can't be undone.",
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'OK',
          onPress: () => {
            this.props.deletePayment(this.props.payment.id)
            this.props.navigation.goBack()
          },
        },
      ]
    )
  }

  render() {
    return (
      <Button
          onPress={() => this.deletePayment()}
          title='DELETE'
          color={DANGER_BUTTON_COLOR}
          accessibilityLabel='Save payment'/>
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    deletePayment: paymentId => {
      dispatch(actions.deletePayment(paymentId))
    }
  }
}

export default connect((state) => ({}), mapDispatchToProps)(DeletePaymentButton)
