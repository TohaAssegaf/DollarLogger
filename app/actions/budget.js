import {
  SET_BUDGET_TOTAL_REQUEST,
  SET_BUDGET_TOTAL_SUCCESS,
  SET_BUDGET_TOTAL_FAILURE,
  GET_BUDGET_TOTAL_REQUEST,
  GET_BUDGET_TOTAL_SUCCESS,
  GET_BUDGET_TOTAL_FAILURE,
  SetBudgetTotalRequestAction,
  SetBudgetTotalSuccessAction,
  SetBudgetTotalFailureAction,
  GetBudgetTotalRequestAction,
  GetBudgetTotalSuccessAction,
  GetBudgetTotalFailureAction
} from './ActionTypes'
import * as BudgetModel from '/app/store/models/budget'

export function setBudgetTotalRequest(): SetBudgetTotalRequestAction {
  return {
    type: SET_BUDGET_TOTAL_REQUEST
  }
}

export function setBudgetTotalSuccess(total: number): SetBudgetTotalSuccessAction {
  return {
    type: SET_BUDGET_TOTAL_SUCCESS,
    total
  }
}

export function setBudgetTotalFailure(errorMessage: string): SetBudgetTotalFailureAction {
  return {
    type: SET_BUDGET_TOTAL_FAILURE,
    errorMessage
  }
}

export function setBudgetTotal(total: number) {
  return function(dispatch) {
    if (total <= 0) {
      dispatch(setBudgetTotalFailure("Budget must be greater than 0"))
    }
    if (!Number.isInteger(total)) {
      dispatch(setBudgetTotalFailure("Budget must be an integer"))
    }
    dispatch(setBudgetTotalRequest(total))
    return BudgetModel.setTotal(total)
      .then(() => {
        dispatch(setBudgetTotalSuccess(total))
        dispatch(NavigationActions.back())
      })
      .catch(error => dispatch(setBudgetTotalFailure(error.message)))
  }
}

export function getBudgetTotalRequest(): GetBudgetTotalRequestAction {
  return {
    type: GET_BUDGET_TOTAL_REQUEST
  }
}

export function getBudgetTotalSuccess(total: ?number): GetBudgetTotalSuccessAction {
  return {
    type: GET_BUDGET_TOTAL_SUCCESS,
    total
  }
}

export function getBudgetTotalFailure(errorMessage: string): GetBudgetTotalFailureAction {
  return {
    type: GET_BUDGET_TOTAL_FAILURE,
    errorMessage
  }
}

export function getBudgetTotal() {
  return function(dispatch) {
    dispatch(getBudgetTotalRequest())
    return BudgetModel.getTotal()
      .then(total => {
        dispatch(getBudgetTotalSuccess(total))
      })
      .catch(error => dispatch(getBudgetTotalFailure(error.message)))
  }
}
