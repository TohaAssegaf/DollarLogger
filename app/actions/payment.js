import {
  CREATE_PAYMENT_REQUEST,
  CREATE_PAYMENT_SUCCESS,
  CREATE_PAYMENT_FAILURE,
  CreatePaymentRequestAction,
  CreatePaymentSuccessAction,
  CreatePaymentFailureAction
} from './ActionTypes'
import * as PaymentModel from '/app/store/models/payment'
import {
  Payment
} from '/app/store/state/PaymentState'

export function createPaymentRequest(): CreatePaymentRequestAction {
  return {
    type: CREATE_PAYMENT_REQUEST
  }
}

export function createPaymentSuccess(payment: Payment): CreatePaymentSuccessAction {
  return {
    type: CREATE_PAYMENT_SUCCESS,
    payment
  }
}

export function createPaymentFailure(errorMessage: string): CreatePaymentFailureAction {
  return {
    type: CREATE_PAYMENT_FAILURE,
    errorMessage
  }
}

export function createPayment(total: number, name: string, date: Date) {
  return function (dispatch) {
    if (!Number.isInteger(total)) {
      dispatch(createPaymentFailure("Payment total must be an integer"))
    }
    dispatch(createPaymentRequest())
    return PaymentModel.createPayment(total, name, date)
      .then(payment => dispatch(createPaymentSuccess(payment)))
      .catch(error => dispatch(createPaymentFailure(error.message)))
  }
}