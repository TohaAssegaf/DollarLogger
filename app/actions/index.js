import * as budget from './budget'

export default Object.assign({}, budget)
