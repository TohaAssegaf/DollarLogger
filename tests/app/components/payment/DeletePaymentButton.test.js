import actions from '/app/actions'
import DeletePaymentButton from '/app/components/payment/DeletePaymentButton'
import * as PaymentModel from '/app/store/models/payment'
import enzyme from 'enzyme'
import { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import configureMockStore from 'redux-mock-store'
import { Alert } from 'react-native'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'

enzyme.configure({ adapter: new Adapter() })
const navigation = { goBack: jest.fn() }
Alert.alert = jest.fn()
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

let mockPayments = []
jest.mock('../../../../app/store/models/payment', () => {
  return {
    deletePayment: jest.fn((item, paymentId) => {
      return new Promise((resolve, reject) => {
        mockPayments = mockPayments.filter(mockPayment => mockPayment.id !== paymentId)
        resolve(mockPayments)
      })
    })
  }
})

it('renders correctly', () => {
  const payment = {
    id: 1,
    total: 1000,
    name: "Test payment",
    date: new Date(2018, 4, 3)
  }
  const store = mockStore({ payment: { payments: [payment] } })
  const rendered = renderer.create(<DeletePaymentButton
    payment={payment}
    store={store}
    navigation={navigation}
  />).toJSON();
  expect(rendered).toMatchSnapshot();
});

it('dispatches delete payment action and navigates back', () => {
  const id = 1
  const total = 10000
  const name = "Test payment"
  const date = new Date(2018, 4, 2)
  const payment = { id, total, name, date }
  const store = mockStore({ payment: { payments: [payment] } })
  const wrapper = shallow(
    <DeletePaymentButton
      payment={payment}
      store={store}
      navigation={navigation}
    />).dive({ context: { store } })

  wrapper.find('Button').simulate('press')

  // Simulate pressing OK on Alert.
  Alert.alert.mock.calls[0][2][1].onPress()

  expect(store.getActions()).toContainEqual(actions.getPaymentsRequest())
  expect(PaymentModel.deletePayment.mock.calls).toEqual([[id]])
  expect(navigation.goBack.mock.calls).toHaveLength(1)
});
