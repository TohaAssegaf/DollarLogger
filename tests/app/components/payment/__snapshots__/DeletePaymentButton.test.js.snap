// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "marginHorizontal": 20,
    }
  }
>
  <View
    accessibilityLabel="Save payment"
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#F44336",
            },
          ]
        }
      >
        DELETE
      </Text>
    </View>
  </View>
</View>
`;
